lightning_pose/metrics.py

```diff
@@ -0,0 +1,184 @@
+
+    # resize reprojections back to original dims
+    keypoints_reproj = _resize_keypoints(cfg, keypoints_reproj_resize, orig_to_resize=False)
+
+    # compute pixel error
+    error_pca = pixel_error(
+        keypoints_pred[:, pca_cols, :].cpu().numpy(), keypoints_reproj.cpu().numpy())
+
+    # next, put this back into a full keypoints pred arr
+    error_all = np.nan * np.zeros((original_dims[0], original_dims[1]))
+    error_all[:, pca_cols] = error_pca
+
+    return error_all
+
+
+@typechecked
+def pca_multiview_reprojection_error(
+    keypoints_pred: Union[np.ndarray, torch.Tensor],
+    pca: KeypointPCA,
+    cfg: DictConfig,
+) -> np.ndarray:
+    """PCA reprojection error.
+
+    Args:
+        keypoints_pred: shape (samples, n_keypoints, 2)
+        pca: pca object that contains info about pca subspace
+        cfg: standard config file that carries around dataset info
+
+    Returns:
+        shape (samples, n_keypoints)
+
+    """
+    from lightning_pose.utils.pca import format_multiview_data_for_pca
+
+    if not isinstance(keypoints_pred, torch.Tensor):
+        keypoints_pred = torch.tensor(keypoints_pred, device=pca.device, dtype=torch.float32)
+    original_dims = keypoints_pred.shape
+
+    mirrored_column_matches = list(pca.mirrored_column_matches)
+
+    # resize predictions to training dims (where pca is computed)
+    keypoints_pred_resize = _resize_keypoints(
+        cfg, keypoints_pred=keypoints_pred, orig_to_resize=True)
+
+    # compute reprojection
+    # adding a reshaping below the loss class expects a single last dim with num_keypoints * 2
+    data_arr = pca._format_data(
+        data_arr=keypoints_pred_resize.reshape(keypoints_pred_resize.shape[0], -1))
+    reproj = pca.reproject(data_arr=data_arr)
+    keypoints_reproj_resize = reproj.reshape(reproj.shape[0], reproj.shape[1] // 2, 2)
+
+    # resize reprojections back to original dims
+    keypoints_reproj = _resize_keypoints(cfg, keypoints_reproj_resize, orig_to_resize=False)
+
+    # put original keypoints in same format
+    keypoints_pred_reformat = pca._format_data(
+        data_arr=keypoints_pred.reshape(keypoints_pred.shape[0], -1))
+    keypoints_pred_reformat = keypoints_pred_reformat.reshape(
+        keypoints_pred_reformat.shape[0], keypoints_pred_reformat.shape[1] // 2, 2)
+
+    # compute pixel error
+    error_pca = pixel_error(keypoints_pred_reformat.cpu().numpy(), keypoints_reproj.cpu().numpy())
+
+    # next, put this back into a full keypoints pred arr
+    error_pca = error_pca.reshape(
+        -1,
+        len(mirrored_column_matches[0]),
+        len(mirrored_column_matches),
+    )  # batch X num_used_keypoints X num_views
+    error_all = np.nan * np.zeros((original_dims[0], original_dims[1]))
+    for c, cols in enumerate(mirrored_column_matches):
+        error_all[:, cols] = error_pca[:, :, c]  # just the columns belonging to view c
+
+    return error_all
```

Comment:
I don't know if a priority but the two losses are very similar at least initially. wondering if we can reuse code in some way. maybe not too urgent

Reply:
good idea, done